subprojects/execution/src/main/java/org/gradle/internal/execution/timeout/impl/DefaultTimeoutHandler.java

```diff
@@ -16,62 +16,118 @@
+
+        private boolean slowStop;
+        private boolean stopped;
+        private boolean interrupted;
+        private ScheduledFuture<?> scheduledFuture;
 
-        private DefaultTimeout(ScheduledFuture<?> timeoutTask, InterruptOnTimeout interrupter) {
-            this.timeoutTask = timeoutTask;
-            this.interrupter = interrupter;
+        private DefaultTimeout(Thread thread, Duration timeout, Describable workUnitDescription, @Nullable BuildOperationRef buildOperationRef) {
+            this.thread = thread;
+            this.timeout = timeout;
+            this.workUnitDescription = workUnitDescription;
+            this.buildOperationRef = buildOperationRef;
+
+            scheduledFuture = executor.schedule(this::interrupt, timeout.toMillis(), TimeUnit.MILLISECONDS);
         }
 
-        @Override
-        public boolean stop() {
-            timeoutTask.cancel(true);
-            return interrupter.interrupted;
+        private void interrupt() {
+            synchronized (lock) {
+                if (!stopped) {
+                    interrupted = true;
+                    doAsPartOfBuildOperation(() -> LOGGER.warn("Requesting stop of " + workUnitDescription.getDisplayName() + " as it has exceeded its configured timeout of " + TimeFormatting.formatDurationTerse(timeout.toMillis()) + "."));
+                    thread.interrupt();
+                    scheduledFuture = executor.schedule(this::warnIfNotStopped, warnIfNotStoppedFrequency(), TimeUnit.MILLISECONDS);
+                }
+            }
         }
-    }
 
-    private static class InterruptOnTimeout implements Runnable {
-        private final Thread thread;
-        private volatile boolean interrupted;
+        private void warnIfNotStopped() {
+            synchronized (lock) {
+                if (!stopped) {
+                    slowStop = true;
+                    doAsPartOfBuildOperation(() -> LOGGER.warn("Timed out " + workUnitDescription.getDisplayName() + " has not yet stopped."));
+                    scheduledFuture = executor.schedule(this::warnIfNotStopped, warnIfNotStoppedFrequency(), TimeUnit.MILLISECONDS);
+                }
+            }
+        }
 
-        private InterruptOnTimeout(Thread thread) {
-            this.thread = thread;
+        private void doAsPartOfBuildOperation(Runnable runnable) {
+            BuildOperationRef previousBuildOperationRef = currentBuildOperationRef.get();
+            try {
+                currentBuildOperationRef.set(this.buildOperationRef);
+                runnable.run();
+            } finally {
+                currentBuildOperationRef.set(previousBuildOperationRef);
+            }
         }
 
         @Override
-        public void run() {
-            interrupted = true;
-            thread.interrupt();
+        public boolean stop() {
+            synchronized (lock) {
+                scheduledFuture.cancel(true);
+                stopped = true;
+                if (slowStop) {
```

Comment:
I ended up not doing that because it feels a little cumbersome/unpolished in the common case where things do stop in time. I think the implemented strategy is a better user experience, though that's highly subjective.